lib/standard/collection/abstract_collection.nit

```diff
@@ -497,6 +497,25 @@ interface MapRead[K, V]
 		end
 		return true
 	end
+
+	# A hashcode based on the hashcode of the keys and the values.
+	#
+	# ~~~
+	# var a = new HashMap[String, Int]
+	# var b = new ArrayMap[Object, Numeric]
+	# a["one"] = 1
+	# b["one"] = 1
+	# assert a.hash == b.hash
+	# ~~~
+	redef fun hash
+	do
+		var res = length
+		for k, v in self do
+			if k != null then res += k.hash * 7
+			if v != null then res += v.hash * 11
+		end
```

Comment:
the semantic of maps is defined in Map (that make sense) and the implementation in the implementation classes (that also make sense).

I did chose that `==` belong to the semantic, so that two maps that associate the same keys with the same values (for a `==` definition of same) are equals.
It is the same with Sequence and Set, and I think it is good.

For instance, in Ruby set are not `==` if they contain the same elements and this make the implementation of some algorithm a PITA.

Now, the law on `==` make that two equals thing **must** have the same hashcode. And because the equality is defined in Map, I thus have no choice to define hashcode in the same class or else I will broke the law (I maybe go to prison or something).